### PR TITLE
8303526: Changing "arbitrary" Name.compareTo() ordering breaks the regression suite

### PR DESCRIPTION
**Pull request by archiecobbs**
This was discovered during investigation for [JDK-8268622](https://bugs.openjdk.org/browse/JDK-8268622).

The method `Name.compareTo()` looks like this:
```java
/** An arbitrary but consistent complete order among all Names.
 */
public int compareTo(Name other) {
    return other.getIndex() - this.getIndex();
}
```
Note the "arbitrary but consistent" wording. Unfortunately, the ordering can be "arbitrary" only if you don't mind breaking the regression suite.

For example, if you apply this patch:
```diff
--- a/src/jdk.compiler/share/classes/com/sun/tools/javac/util/Name.java
+++ b/src/jdk.compiler/share/classes/com/sun/tools/javac/util/Name.java
@@ -105,7 +105,7 @@ public abstract class Name implements javax.lang.model.element.Name, PoolConstan
     /** An arbitrary but consistent complete order among all Names.
      */
     public int compareTo(Name other) {
-        return other.getIndex() - this.getIndex();
+        return this.getIndex() - other.getIndex();
     }

     /** Return true if this is the empty name.
```
then the following tests will fail:
```TEST: tools/javac/8203436/T8203436b.java
TEST: tools/javac/generics/7034019/T7034019c.java
TEST: tools/javac/generics/7034019/T7034019d.java
TEST: tools/javac/generics/diamond/neg/Neg21.java
TEST: tools/javac/generics/diamond/neg/Neg22.java
TEST: tools/javac/generics/inference/EagerReturnTypeResolution/EagerReturnTypeResolutionTestb.java
TEST: tools/javac/patterns/InferenceUnitTest.java
TEST: tools/javac/T8187978/FilterOutCandidatesForDiagnosticsTest.java
TEST: tools/javac/varargs/6806876/T6806876.java

==============================
Test summary
==============================
   TEST                                              TOTAL  PASS  FAIL ERROR
>> jtreg:test/langtools:langtools_javac               3718  3709     9     0 <<
==============================
```
The compiler behavior should be deterministic whenever possible, and the javac regression suite should be robust. So clearly "arbitrary" behavior by this method is not really supported.

It turns out there is only one user of `Name.compareTo()`, the method `TypeSymbol.precedes()`:
```java
/**
 * A partial ordering between type symbols that refines the
 * class inheritance graph.
 *
 * Type variables always precede other kinds of symbols.
 */
public final boolean precedes(TypeSymbol that, Types types) {
    if (this == that)
        return false;
    if (type.hasTag(that.type.getTag())) {
        if (type.hasTag(CLASS)) {
            return
                types.rank(that.type) < types.rank(this.type) ||
                types.rank(that.type) == types.rank(this.type) &&
                that.getQualifiedName().compareTo(this.getQualifiedName()) < 0;
        } else if (type.hasTag(TYPEVAR)) {
            return types.isSubtype(this.type, that.type);
        }
    }
    return type.hasTag(TYPEVAR);
}
```
If we want the above partial ordering to be stable, we need to specify deterministic behavior for `Name.compareTo()`.

An obvious choice would be to require that `Name.compareTo()` orders consistently with `String.compareTo()`, i.e., lexicographically on the Unicode characters in the name.

Fortunately, this comparison can be done efficiently, i.e., without actually converting the `Name` into a `String`, because UTF-8 is "lexicographically consistent" with the characters it encodes, except that Java's use of Modified UTF-8 means a special check for 0x0000 (which is encoded as two bytes) is required.

When you do this the tests are still broken... but notice that `TypeSymbol.precedes()` is actually using the _reverse_ of `Name.compareTo()`. So somehow, the current `Name.compareTo()` is ordering things "backwards", at least from the point of view of the expectation of the regression tests.

So replacing `Name.compareTo()` with a version that orders lexicographically, and also un-reversing its treatment in `TypeSymbol.precedes()`, makes things the tests succeed again.

That's what this patch does.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8303526](https://bugs.openjdk.org/browse/JDK-8303526): Changing "arbitrary" Name.compareTo() ordering breaks the regression suite

### Reviewers
 * [Vicente Romero](https://openjdk.org/census#vromero) (@vicente-romero-oracle - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/12843/head:pull/12843` \
`$ git checkout pull/12843`

Update a local copy of the PR: \
`$ git checkout pull/12843` \
`$ git pull https://git.openjdk.org/jdk.git pull/12843/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12843`

View PR using the GUI difftool: \
`$ git pr show -t 12843`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12843.diff">https://git.openjdk.org/jdk/pull/12843.diff</a>

</details>

### CODE REVIEW

File: src/jdk.compiler/share/classes/com/sun/tools/javac/util/Name.java [hunk 1]
```diff
@@ -38,7 +38,7 @@
  *  This code and its internal interfaces are subject to change or
  *  deletion without notice.</b>
  */
-public abstract class Name implements javax.lang.model.element.Name, PoolConstant {
+public abstract class Name implements javax.lang.model.element.Name, PoolConstant, Comparable<Name> {
```
Comment (jonathan-gibbons) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/util/Name.java:41]: It's disappointing that we can't easily make `javax.lang.model.element.Name` implement `Comparable`

**Review by jonathan-gibbons — COMMENTED** [review]
Looks OK to me, but I'll stop short of the "A" word so that someone on the javac team can review it as well.

**Comment by archiecobbs** [comment]
> Looks OK to me, but I'll stop short of the "A" word so that someone on the javac team can review it as well.

Thanks for taking a look. I'll promote from draft & we'll see what other folks think.

**Comment by vicente-romero-oracle** [comment]
not totally sure what we are fixing here, I mean it could be that the current code is not arbitrarily consistent but it is faster than the proposed implementation

**Comment by archiecobbs** [comment]
> sorry if I'm missing something but I'm not totally sure what we are fixing here, I mean it could be that the current code is not arbitrarily consistent but it is faster than the proposed implementation

I can explain the problem we're fixing... I think the real question here though is whether the problem we're fixing is a big enough problem that it need to be fixed, or just not worth worrying about.

The problem is that the specification of `Name.compareTo()` is so loose that it introduces a source of non-determinism in the compiler that is dependent on (for example) the order in which class files are read. That non-determinism then flows into `TypeSymbol.precedes()`.

So the real question is, do we really want `TypeSymbol.precedes()` potentially sorting `TypeSymbols` differently each time?

I can't really say definitively, because I'm not familiar with this method - so I'm just basing my opinion on the circumstantial evidence that there are 9 regression tests that are sensitive to this ordering, and also my overall general feeling that needless non-determinism is always a bad thing in computer software because it makes everything harder to debug.

The downside of this patch is a slower implementation of `TypeSymbol.precedes()` - again, I'm not familiar with this method so unsure how critical it is. Also, note that this downside may happen in the future regardless if/when the `Name` classes get refactored to just be `String` wrappers, etc.

So... I dunno. Your thoughts?

**Comment by vicente-romero-oracle** [comment]
> > sorry if I'm missing something but I'm not totally sure what we are fixing here, I mean it could be that the current code is not arbitrarily consistent but it is faster than the proposed implementation
> 
> I can explain the problem we're fixing... I think the real question here though is whether the problem we're fixing is a big enough problem that it need to be fixed, or just not worth worrying about.
> 
> The problem is that the specification of `Name.compareTo()` is so loose that it introduces a source of non-determinism in the compiler that is dependent on (for example) the order in which class files are read. That non-determinism then flows into `TypeSymbol.precedes()`.
> 
> So the real question is, do we really want `TypeSymbol.precedes()` potentially sorting `TypeSymbols` differently each time?
> 
> I can't really say definitively, because I'm not familiar with this method - so I'm just basing my opinion on the circumstantial evidence that there are 9 regression tests that are sensitive to this ordering, and also my overall general feeling that needless non-determinism is always a bad thing in computer software because it makes everything harder to debug.
> 
> The downside of this patch is a slower implementation of `TypeSymbol.precedes()` - again, I'm not familiar with this method so unsure how critical it is. Also, note that this downside may happen in the future regardless if/when the `Name` classes get refactored to just be `String` wrappers, etc.
> 
> So... I dunno. Your thoughts?

I see, patch looks good, we will though have to be on watch for possible performance regressions, but we'll see

**Review by vicente-romero-oracle — APPROVED** [review]
looks good

**Pull request closed without merging**